webcrypto/subtle_crypto.go

```diff
@@ -137,8 +141,87 @@ func (sc *SubtleCrypto) Verify(
 //   - SHA-512
 //
 // The `data` parameter should contain the data to be digested.
-func (sc *SubtleCrypto) Digest(algorithm goja.Value, data interface{}) *goja.Promise {
-	return nil
+func (sc *SubtleCrypto) Digest(algorithm interface{}, data goja.Value) *goja.Promise {
+	promise, resolve, reject := sc.makeHandledPromise()
+
+	// Validate that the value we received is either an ArrayBuffer, TypedArray, or DataView
+	// This uses the technique described in https://github.com/dop251/goja/issues/379#issuecomment-1164441879
+	var (
+		isArrayBuffer = IsInstanceOf(sc.vu.Runtime(), data, ArrayBufferConstructor)
+		isDataView    = IsInstanceOf(sc.vu.Runtime(), data, DataViewConstructor)
+		isTypedArray  = IsTypedArray(sc.vu.Runtime(), data)
+	)
+
+	if !isArrayBuffer && !isDataView && !isTypedArray {
+		reject(fmt.Errorf("data must be an ArrayBuffer, TypedArray, or DataView"))
+		return promise
+	}
+
+	// 2.
+	// Cast the data to a Goja Object, and, as we're now sure it's
+	// either an ArrayBuffer, or a view on an ArrayBuffer, we can
+	// get the underlying ArrayBuffer by exporting its buffer property
+	// to a Goja ArrayBuffer, and then getting its underlying Go slice
+	// by calling the `Bytes()` method.
+	//
+	// Doing so conviniently also copies the underlying buffer, which
+	// is required by the specification.
+	// See https://www.w3.org/TR/WebCryptoAPI/#SubtleCrypto-method-digest
+	asObject := data.ToObject(sc.vu.Runtime())
+	arrayBuffer, ok := asObject.Export().(goja.ArrayBuffer)
+	if !ok {
+		reject(fmt.Errorf("could not get ArrayBuffer from data"))
+		return promise
+	}
+	bytes := arrayBuffer.Bytes()
+
+	// The specification explicitly requires us to copy the underlying
+	// bytes held by the array buffer
+	bytesCopy := make([]byte, len(bytes))
+	copy(bytesCopy, bytes)
+
+	// 3.
+	normalizedAlgorithm, err := NormalizeAlgorithm(algorithm, OperationIdentifierDigest)
+	if err != nil {
+		// "if an error occurred, return a Promise rejected with NormalizedAlgorithm"
```

Comment:
:thinking: I spotted this in WebCrypto specification, but to be honest, I'm not sure if I get this comment. Maybe I'm (and imagine like any other reader of this code) missing some context that we can put here.

What does mean `rejected with NormalizedAlgorithm"`?

Reply:
This comment is legacy and reminded me to change the older implementation. I'll drop it.

The Web Crypto specification is very cryptic in certain regards, indeed. It took me a long time and a lot of research to determine what that meant. Most of what I've done on that front is inspired by the [Deno implementation](https://github.com/denoland/deno/blob/main/ext/crypto/00_crypto.js). 

The algorithm normalization is essentially the step of converting an input `Object` or `string`, in a predefined format, into a (set of) internal structures that can be worked on while also ensuring to fail if any validation steps fail. It's a step that's quite easy in a dynamic language but was quite hard to do in a static language. I'm still not happy with the results, but it will do for now :)

I'll add a comment here to explain just that 👍🏻

Reply:
I feel like this comment *made* it worse :rofl: 

The specification says:
```
Let normalizedAlgorithm be the result of [normalizing an algorithm](https://www.w3.org/TR/WebCryptoAPI/#dfn-normalize-an-algorithm), with alg set to algorithm and op set to "digest".

If an error occurred, return a Promise rejected with normalizedAlgorithm.
```

And as the Specification does not have multiple returns `normalizedAlgorithm` is just the result of the operation which *can* be an error (which "occurred"). 

So from that I would recommend *not* having a comment as that IMO makes it worse 
```suggestion
```